apis/v1alpha1/gatewayclass_types.go

```diff
@@ -64,23 +64,13 @@ type GatewayClassSpec struct {
 	Controller string `json:"controller"`
 
 	// AllowedGatewayNamespaceSelector is a selector of namespaces that Gateways
-	// can use this GatewayClass from. This is a standard Kubernetes
-	// LabelSelector. Controllers must not support Gateways in namespaces
-	// outside this selector.
-	//
-	// An empty selector (default) indicates that Gateways can use this
-	// GatewayClass from any namespace.
-	//
-	// When a Gateway attempts to use this class from a namespace that is not
-	// allowed by this selector, the controller implementing the GatewayClass
-	// may add a new "ForbiddenNamespaceForClass" condition to the Gateway
-	// status. Adding this condition is considered optional since not all
-	// controllers will have access to all namespaces.
+	// of this class can be created in. Gateway validation will prevent Gateways
```

Comment:
Can we change validation to validation webhook to make it more specific?
Other than that this lgtm.

Reply:
Isn't it frowned upon to have validation cross-resource since it can invalidate objects already persisted, cause ordering issues, etc? Or is that just an Istio practice not a general Kubernetes practice?

Reply:
Thanks for catching that @hbagdi! I'll change the wording to include "webhook". I completely agree that any kind of cross resource validation is less than ideal and generally something that should be avoided. Unfortunately I haven't been able to come up with a better alternative here. I think this is the least bad option available, but I'm open to other options.

This PR intentionally changes the behavior of this field to restrict creation of Gateways, not updates. It intentionally will not invalidate existing Gateways if this field changes. Before this PR, changes to this field could have quite destructive results by invalidating all Gateways of a class. Essentially this shifts the burden of cross resource validation from controllers to a webhook and moves it from validation required on every Gateway or GatewayClass change to something that happens once at Gateway creation.

I think the distinction of this validation only applying on create and the expectation that this field on GatewayClass will not change very frequently makes the cross resource validation an acceptable compromise here. Unfortunately I don't know of any way to accomplish this with anything other than OPA which would end up having similar issues.

Reply:
Should we reach out to some of the API machinery folks to see if there are some recommended practices here? The Istio community has already come the conclusion we will never support cross resource validation webhooks, so it seems unlikely we can support this part of the spec (which is fine if that's what the project wants, we are just one implementor).

I guess what I am trying to get at is cross resource validation a "bad practice" or a "really really bad idea". If it's the former and there are no alternatives it seems reasonable. If it's the latter, which is my personal opinion, then even if there are not alternatives we still should not do it.

Reply:
You're completely right, asked in Slack and will share any answers I get here. If I don't get anything I'll reach out to someone directly tomorrow.

Reply:
@robscott any updates regarding alternatives to cross resource validation?

Reply:
I addressed this in my [comment below](https://github.com/kubernetes-sigs/service-apis/pull/263#issuecomment-690821018) but should have followed up here as well. I [asked in Slack](https://kubernetes.slack.com/archives/C0EN96KUY/p1599699017096700) and the general theme of the answer was not that we can't but that we probably shouldn't. Here are a couple of the tips from Jordan:

>admission that checks other objects to try to enforce uniqueness of an attribute is always racy with concurrent requests

> instead of synchronous protection like that, I'd encourage having good reconciliation and status reporting of conflicts which can handle all cases

With that in mind, I've [proposed yet another variation](https://github.com/kubernetes-sigs/service-apis/pull/263#issuecomment-690821018) below that is a bit of a hybrid approach.